### DISCUSSION TITLE
How to make ripgrep match multiple rules in one line or use longest match principle？

### DISCUSSION THREAD

**Discussion by XhstormR** [BurntSushi/ripgrep#1644] [General] [answered]
This is the content of the data.txt:
```
www.yxlink.com/
www.china.com/
www.goodfood.com/
```

This is the content of the rule.txt:
```
(?-u:www.)
(?-u:www.yxlink.com/)
(?-u:www.china.com/)
(?-u:www.goodfood.com/)
```

This is the command:
```
rg -f rule.txt data.txt
```

This is the result after execution:
![1](https://user-images.githubusercontent.com/10527522/88273803-50aceb00-cd0d-11ea-8bfd-7866d4c92bc0.jpg)

You can see that only the first rule was matched successfully, and all the following rules failed to match.

When I remove the first rule `(?-u:www.)`, it can be matched successfully:
![2](https://user-images.githubusercontent.com/10527522/88274339-21e34480-cd0e-11ea-9b12-a7038ce40680.jpg)

How to make ripgrep match multiple rules in one line or use longest match principle？

**Comment by BurntSushi** [comment, marked as answer]
The only way is to sort your rules file by the length of each rule in descending order.

**Reply by BurntSushi** [reply]
For further reading, please see my `aho-corasick` docs on match kind: https://docs.rs/aho-corasick/0.7.13/aho_corasick/enum.MatchKind.html --- It's probably the best I've written on the topic.

If you have a specialized task where all you need to do is search for a fixed set of non-regex strings, then it would be very simple to write a Rust program that uses the `aho-corasick` crate and enables leftmost longest match semantics. Then you wouldn't need to sort your rules, if that matters.

I do feel obligated to note though that your existing setup _is_ interpreted your rules as regexes. In particular, the `.` is a special meta-character that matches any character, and you probably want to have it match a literal `.`. You can fix this by either escaping the `.` via `\.`, or by passing the `-F` flag to ripgrep, which will treat the rules as literal non-regex patterns.

**Comment by XhstormR** [comment]
Thank you very much for your help!
